control/rebalance.py

```diff
@@ -0,0 +1,162 @@
+#
+#  Copyright (c) 2024 International Business Machines
+#  All rights reserved.
+#
+#  SPDX-License-Identifier: LGPL-3.0-or-later
+#
+#  Authors: leonidc@il.ibm.com
+#
+
+#import uuid
+import errno
+import threading
+import time
+import json
+import re
+from .utils import GatewayLogger
+from .config import GatewayConfig
+from .proto import gateway_pb2 as pb2
+
+class Rebalance:
+    """Miscellaneous functions which do rebalance of ANA groups
+    """
+
+    def __init__(self, gateway_service):
+        self.logger = gateway_service.logger
+        self.gw_srv = gateway_service
+        self.ceph_utils = gateway_service.ceph_utils
+        self.rebalance_period_sec              = gateway_service.config.getint_with_default("gateway",  "rebalance_period_sec", 7)
+        self.rebalance_max_ns_to_change_lb_grp = gateway_service.config.getint_with_default("gateway", "max_ns_to_change_lb_grp", 8)
+        self.auto_rebalance = threading.Thread(target=self.auto_rebalance_task, daemon=True)
+        self.auto_rebalance.start() #start the thread
+
+    def auto_rebalance_task(self ):
+        """Periodically calls for auto rebalance."""
+        if (self.rebalance_period_sec > 0):
+            while True:
+               for i in range(self.rebalance_max_ns_to_change_lb_grp):
+                  try:
+                     rc = self.gw_srv.execute_grpc_function(self.rebalance_logic, None, "context")
+                     if rc == 1:
+                         self.logger.info(f"Nothing found for rebalance, break at {i} iteration")
+                         break
+                  except AssertionError:
```

Comment:
No need for special handling of assert. This will be caught by the Exception clause. In any case the log function will show you the exact exception.